### PR TITLE
Enfore TableScanNode's currentConstraint is set only when layout chosen

### PR DESCRIPTION
**Pull request by findepi**

### CODE REVIEW

**Comment by findepi** [comment]
cc @martint

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/TableScanNode.java [hunk 1]
```diff
@@ -76,6 +76,7 @@ public TableScanNode(
         checkArgument(assignments.keySet().containsAll(outputs), "assignments does not cover all of outputs");
         requireNonNull(tableLayout, "tableLayout is null");
         requireNonNull(currentConstraint, "currentConstraint is null");
+        checkArgument(currentConstraint.isAll() || tableLayout.isPresent(), "currentConstraint present without layout");
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/TableScanNode.java]: why do we need `currentConstraint` at all?  There is `tableLayout.getPredicate()` which in my understaning should match `currentConstraint`.

Also it looks like this field is not used, I only see references to it from rewrite rules.

Comment for this field says: `Used during predicate refinement over multiple passes of predicate pushdown`. However this field is not used in `PredicatePushdown`.

Lets remove this field? 
@martint Have I missed something?

Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/plan/TableScanNode.java]: In `AddExchanges` [i see](https://github.com/prestodb/presto/blob/dcb07d32562e9f616268998324f29c3b295d5c51/presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java#L619):
```
new currentConstraint := (constraint we tried to push down to connector, including old currentConstraint).intersect(layout.getLayout().getPredicate())
```
which suggests `layout.getLayout().getPredicate()` might be weaker (or we don't fully trust connectors?) than the `currentConstraint`. 

Just thinking... I'd rather know for sure from @martint .

Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/plan/TableScanNode.java]: I was also thinking... 

It seems to be used in `EffectivePredicateExtractor` which is used in `PredicatePushdown` run after `AddExchanges`. That means that information set in `currentConstraint` by `AddExchanges` are then used `PredicatePushdown`.

Comment (kokosing) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/plan/TableScanNode.java]: One more thing. Notice that `TableScanNode` has `TableLayoutHandle`, so you cannot invoke `tableLayout.getPredicate() `, you need get `TableLayout` first.

**Pull request merged**